### ISSUE TITLE
General Discussion Thread

### ISSUE THREAD

**Issue by the-m4a** [civfanatics/CQUI_Community-Edition#3]
(Starting this just to cover this question):
How do we get this published to the Workshop?  @alexeyOnGitHub do you own that publishing mechanism?  I recall that Azurency said someone else needed to do the publishing as he could not.

If we need to re-publish it ourselves, we should change the Mod ID GUID in the ModInfo file and list the old (current) GUID as an incompatible mod.

**Comment by alexeyOnGitHub** [comment]
I do not own that part. so far I see people suggesting to use Workshop Uploader in Civilization 6 SDK (which can be installed from Steam). this should work for a new mod. need to check what the deal is with permissions and such. hey @Azurency , can you please share some notes on how you published this mod in the past? I want to keep this info in a file in the repo. maybe `CONTRIBUTING.md` or something similar.

**Comment by the-m4a** [comment]
Anyone else see Spanish (I think?) in the City-State text?
![image](https://user-images.githubusercontent.com/8787640/83595543-23b33780-a517-11ea-9e86-c67dc84a0516.png)

**Comment by alexeyOnGitHub** [comment]
I would submit a separate ticket (Github issue) for this. do you see this for every city state? or just this one? what is the name of the city?

**Comment by the-m4a** [comment]
I need to investigate more but no not every one, appears to be only the new ones?

**Comment by Infixo** [comment]
So, guys, have you thought about what to do with the mod? Like what direction to go, what rules to follow, etc. A bit ot strategic thinking.
I am curious because I'd like to contribute but my involvement depends on the above.

**Comment by the-m4a** [comment]
My thinking is to, at the very least, try to clean it up so that it is easier to update in the future whenever Firaxis sends out an update.  I am thinking that adopting the policy of trying to reduce over-writing the Firaxis-authored versions of the Lua files will help this.  And at least for the time being, it's kind of interesting work, especially to try and re-think some of the implementation strategy to avoid having code that does more work than necessary (your suggested change for the housing calculation is a good example of this).

Once we're at that point, I am not certain, but a few things come to mind:
- Ensuring it behaves well with other mods that are popular and useful (Better Reports, Sukitract's Global Relations panel comes to mind).  The work there is likely figuring out "proper" load order values.
- Similar to above: the stuff that we currently have integrated (like Better Espionage, Trade Screen) - allow the actual mods to be used instead, if they're present.  This is again likely a load order thing... because I don't think there's a way to conditionally load items (or even: evaluate those conditions)
- Enabling/Disabling some of the features of the mod, for example: allow the Gossip/Status messages to have the look they currently do in the unmodified game
- Build Queue improvements, like what is mentioned in issue #38 - queue up items that require other items be built first.  The base game does not do this, and being able to queue up all of the pieces of a district would be a fantastic feature.  It'd require some logic be executed whenever something was added/removed from the queue, or if something in the queue was "next" but didn't have the prerequisites, it gets skipped, etc.
- I'd love to have some way of persisting the settings, but I am thinking there's no reasonable way to accomplish that
- (There's one other thing that popped in to my head earlier but I forgot :D)

**Comment by Infixo** [comment]
Here are a few of my thoughts.
1. I think CQUI grew too big to effectively be maintained. I used it for a long time and with every patch the time to make it operational was longer and longer.
2. Stop making copies of game file for every little change that in the long run is not super useful but makes maintaing the mod harder and harder. Get rid of "integrations" that have minimal value added.
3. CQUI should focus on the main functions like city view, city panel, production queue, world view, top bar, launch bar, etc. Everyhing a user sees all the time and interacts the most. The rest should be left for independent mods, with some rare exceptions maybe.
4. Create a list of compatible mods. Like a "compatibility certificate". If a modder wants a mod to be there, then it should follow some rules like proper load order ensuring that the mod works with CQUI, maybe the use of settings, etc.
5. And last, but not least. Stop changing every tab to 2 spaces. Firaxis uses 4 spaces, and this rule is absurdly stupid.

**Comment by the-m4a** [comment]
```
And last, but not least. Stop changing every tab to 2 spaces. Firaxis uses 4 spaces, and this rule is absurdly stupid.
```
LOL b/c while not a fan of the 2 spaces, it appeared to fit most of what had already existed with the CQUI scripts.  If folks want to go to 4 spaces, I would be in favor of it as everything else I work with is 4 spaces.

I prefer the spaces over tabs b/c I work with whitespace characters enabled, which helps me with line readability, keeping things aligned and whatnot.

![image](https://user-images.githubusercontent.com/8787640/85079286-4f6f2800-b17b-11ea-9e2e-ef37494cc76d.png)

**Comment by the-m4a** [comment]
I also agree with some of your other points, especially in light of only a few of us doing any sort of work.  Completely agree with your 2nd point - copying the entirety of the game file for a small change is frustrating whenever there's updates as it's not always easy to determine what Firaxis changed and what was changed by CQUI.  Still working to get away from the full file replacements.

**Comment by the-m4a** [comment]
sorry - to be clear: I agree with all of your other points.  I don't want to expand the mod, and removing the integrations of existing mods (like, if the Diplomacy Deal one is still maintained) and pointing folks at those mods would be nice

My primary use for CQUI is the build queue stuffs, and how the city screen looks.  when trying to fix it back in April, that was the thing I primarily wanted to work.

**Comment by Infixo** [comment]
Great. It seems that our thinking is aligned so I am willing to contribute.
I'd like you to make me a collaborator - it will make things easier and faster.

**Comment by alimulhuda** [comment]
Just realized there was a conversation going on here. Wanted to add my two cents.

- I am also for removing support for integrations, but only for situations where the alternative is maintained. I think we should also talk about it on a case by case basis before we just remove it.
- Smaller and more focused is better. Extensions are better than full overwrites. Copy/paste then modify should be avoided whenever possible. @the-m4a has been making pretty good progress on this end.
- And lastly about the spaces comment. I agree that 4 spaces is more standard than 2 spaces used by this project. I would generally advice against large project wide changes for the sake of personal preference. I've been on teams and projects where the guidelines for coding standard were different (tabs vs spaces, number of spaces used, variable naming conventions, etc), but the overriding rule was not to change the status quo for the project. Large changes work for individual personal projects but for anything else it just causes headache for others, change trackers usually have issues if two people try to change the same line, so if one changes it for aesthetics and another for functionality then there might be conflict we would have to individually resolve. **All that being said, if you all are really against it then I suggest we do this change sooner rather than later**. Also sorry about the wall of text, any time something comes up that has negatively affected me in the past, I get a bit rambly.
- The other points all sound good, I think the mod could use all the help it can get.

Also @Infixo we have some rules setup for this repo regarding contributing. The process we have been following is generally as follows:
1. Create/modify issue to track change being done (assign an issue to yourself so others know you are working on it and avoid having multiple people doing the same thing)
2. Create branch in this repo for working on that issue.
3. Do a pull request to merge the changes to master branch.
4. Ping the civfanatics-cqui group as reviewers or just individually mention whoever's attention you want for code review.
5. After the code review process when you have 2 approving reviews, merge into master branch.

The code review process does slow things down somewhat but I think the way @alexeyOnGitHub set it makes the changes a bit more robust and solid.

**Comment by alexeyOnGitHub** [comment]
@Infixo  , I sent you an invite to join the dev team for this repo.

**Comment by Infixo** [comment]
@alimulhuda 
So, what’s your approach to 4 vs. 2 spaces? I am not quite sure what you mean by asking if we are against it?

When I mentioned this, it was only to say that whatever code I would write, it would be formatted for 4 spaces, just as Firaxis does. I don’t expect anyone to systematically rewrite the code back from 2 to 4 spaces, that’s just plain waste of time. Also, I am not going to do that, unless e.g. copying some big parts of the original firaxis file because of the patch, etc.

As for the contribution process, I have a question. If I understand correctly - nobody can merge into the master without 2 reviews? And how many people are there now who can do that?

**Comment by alexeyOnGitHub** [comment]
2 approvals are required from people with "WRITE" access to the repo to be able to merge a PR. 
an admin can still merge into "master", ignoring the 2-review requirement. everyone in "civfanatics-cqui" team has "WRITE" permissions on this repo. there are currently 3 people in that team, plus I sent you an invite too @Infixo

**Comment by alimulhuda** [comment]
@Infixo Ahh okay I think i just misunderstood your comment. I thought you were advocating changing the current files into 4 space indentation. If I understand you correctly you are saying we follow a 4 spaces rule for any new files? What would you do if you were modifying an existing file which uses 2-space indentation with new functions or bugfixes?

The one thing I would be against is having mixed indentations in one file. So avoid having 4-spaced and 2-spaced in one file. I'm up for whatever you all decide, ideally project wide consistency, but if not possible then at least consistent on file by file basis.

**Comment by Infixo** [comment]
@alimulhuda 
It really depends on the situation. My time is more valuable for me than having 2 and 4 spaces indents in a file. Also, people don't care how many spaces we use, they care if the mod is operational, bugs are fixed and ideally some cooperation with other mods is working.
So, if I am going to change several lines out of 1000 then I will change those several lines and certainly not the entire file if it has been heavily modded so far. But if the rest can just be copied from the original, then probably I would just copy it.

**Comment by the-m4a** [comment]
on the spaces/tabs: I have been working on the CityBannerManager files this week, trying to get them boiled down to just extension-type files, and used 4 spaces instead of 2.  The readability is much better with the 4 spaces.  I'm happy to go with 4 spaces going forward.

FWIW, Firaxis uses tabs, which leaves the number of spaces up to the editor application in terms of displaying it as 2 spaces or 4 spaces.  I don't know offhand if Notepad++ and VSCode use 2 spaces for tabs with Lua files by default... it could be that the Lua extension I have for VSCode does this.  There's a setting in VSCode, _Detect Indentation_, that when I have it set, the tab character is displayed as 2 spaces.

![image](https://user-images.githubusercontent.com/8787640/85209127-3d5fc780-b2ea-11ea-94bf-ba8db754d833.png)

When it's enabled, it _sometimes_ enforces the 2 characters for tabs, and I haven't figured exactly what causes that, but either way:
![image](https://user-images.githubusercontent.com/8787640/85209189-c2e37780-b2ea-11ea-8a8d-4c1cfba5420e.png)

When it's disabled, it honors the 4-spaces for the tab character:
![image](https://user-images.githubusercontent.com/8787640/85209001-7ba8b700-b2e9-11ea-8f3a-f221bf8a5326.png)

**Comment by wayneb64** [comment]
First let me acknowledge my opinion has zero weight as I am not actually working on this but it reminds me of a time almost 20 years ago when I was working at a software company. There was a raging debate about tabs versus spaces among the developers. I was in the tabs camp for a few reasons (less typing and variable width so those who prefer 2, 3, or more spaces could adjust with their editor) and one of the guys in the other camp was so adamant about it  he went around one morning and stole everyone's tab keys and taped them to his door. I don't remember which side on in the end but it seems like it was tabs because I am still using them religiously today.

**Comment by the-m4a** [comment]
Yeah - have had similar debates at work. There it was largely because the (PR) code reviews would show 8(!) spaces per tab when viewed in the browser,  so it was a mess.  I personally like the spaces entirely because of the dotted lines on screen, instead of the -> symbol, as I prefer working with white space characters visible.

The tab key inserting 4 spaces (instead of actual \t character) setting is one I always set as well, so my tab key does get plenty of use :)

**Comment by alimulhuda** [comment]
Tabs vs spaces always leads to insane "debates" (usually just both camps sticking to their guns), which is why my personal policy is just to be consistent. If I were working on a new project by myself (only my tools and zero collaboration) then I would also do what wayneb64 does and use tabs and also set my viewer to show tab characters with size 4.

However, in group projects I think we need to stick to some conventions and I'm okay with having the granularity down to a file level scope, but no finer. The point of whitespace in a language like Lua is readability, and having inconsistent spacing screws with that. Most editors I've used for code (Visual Studio, Visual Studio Code, Notepad++) allow for a quick way to switch the amount/type of spacing used for indentation. As an aside, if you use a editor which honours the .editorconfig file then you don't even have to change your settings, it will auto-adjust based on the .editorconfig file.

> Also, people don't care how many spaces we use, they care if the mod is operational, bugs are fixed and ideally some cooperation with other mods is working.

I agree that users of the mod don't care what we do internally. They also don't care about internal details like whether or not we copy files instead of extending them, but we still want to do it because it makes life easier for future development. The whitespace guidelines we agree to are not for users but for developers (current and future).

**Comment by Infixo** [comment]
@the-m4a Originally TAB code was used to represent a tabulation of 8 columns.

Anyway, can we conclude it like this?
1. Indenting of 4 columns done using either a TAB (then set your editors for 4 spaces) or plain 4 spaces, whatever suits you. Basically the same style as Firaxis uses.
2. If you have an opportunity and are willing to put some time into it, then gradually convert the files to their original formatting, but that is not a priority neither a requirement atm.

**Comment by Infixo** [comment]
Some other topic, while talking about about standards. I assume that changing file names to lowercase has something to do with Linux. Can anyone of you confirm that this is the case and CQUI runs on Linux as of now?

**Comment by alimulhuda** [comment]
> @the-m4a Originally TAB code was used to represent a tabulation of 8 columns.
> 
> Anyway, can we conclude it like this?
> 
>     1. Indenting of 4 columns done using either a TAB (then set your editors for 4 spaces) or plain 4 spaces, whatever suits you. Basically the same style as Firaxis uses.
I'm good with the 4 spaces rule. I would say no to TAB character in files.

>     2. If you have an opportunity and are willing to put some time into it, then gradually convert the files to their original formatting, but that is not a priority neither a requirement atm.
Agreed
From what was talked about before I'm assuming @the-m4a is also in agreement. So unless @alexeyOnGitHub has any objections we can just modify the CONTRIBUTING.md file and start following that.

Regarding the lowercase thing, I don't know why it was done but it seems that people on Linux still have to go through some hoops to get the mod to work. There is an issue #36 where someone posted a workaround that is required for CQUI to work in Linux. The problem is neither @the-m4a or I play on Linux, so I'm not sure how to go about tackling that one.

**Comment by Infixo** [comment]
> I'm good with the 4 spaces rule. I would say no to TAB character in files.

I am using Notepad++ and I don't see an option to somehow autoconvert tabs to spaces. If you know how to do that then let me know pls. As for now for me it is a no go. The whole idea of using a tab is to save you clicks, not to add more and count spaces on top of that.

**Comment by alimulhuda** [comment]
### Instructions for inserting spaces instead of tab character
In Notepad++ go to Settings->Preferences...

![image](https://user-images.githubusercontent.com/178540/85214306-019f1f00-b337-11ea-8549-fbab5b4b9f7b.png)

This setting is where you can tell it to insert spaces instead of tabs whenever you press the tab key on the keyboard. If the "Replace by space" is turned on it will insert the correct number of spaces to align to the next tab stop location. So lets say you insert a space then you press tab then it will insert the remaining number of spaces (1 if tab size is set to 2 and 3 if tabsize is set to 4). If you just press tab on an empty line it will just insert the required number of spaces.

### Instructions for updating full file into one indentation style
![image](https://user-images.githubusercontent.com/178540/85214653-dbc84900-b33b-11ea-9c96-25841b8d2274.png)

If going from tab character to spaces. Set the desired tab size from the preferences section earlier.
Then use Edit->Blank Operations->TAB to Space

If going from spaces to tab characters. First set the tab size to the size used in the file (For the CQUI files right now that would be 2). Then use Edit->Blank Operations->Space to TAB (Leading). Then set the tab size to what you want in the preferences.

@Infixo Let me know if you were asking about something else in your last comment. I'm on version 7.8.6 of Notepad++. If you are on a different version and the options aren't in the same place let me know.

**Comment by the-m4a** [comment]
I man in agreement.  Wouldn’t mind converting things to 4 space indent and whatnot.  I will use that kind of spacing going forward

**Comment by Infixo** [comment]
@alimulhuda Thank you for those instructions, I appreciate.

**Comment by Infixo** [comment]
Guys, I have read your comments for the fixes I did yesterday and I am afraid that our goals and priorities are not aligned in the end.
I really don't wanna spend my time explaining why I from time to time use a single line vs. multiple (there is no such rule in Lua), or maybe why I put a semicolon even when it is not needed (I do because I grew up on C/C++), or why I don't use brackets between if-then (not needed), or if there should be an empty line after end or not. And there is still a matter of the Hungarian notation, etc. And we didn't even get to really important things like code structure, reusability, objects, and new functionalities. All these just from several lines of code.
It is not fun.

**Comment by Infixo** [comment]
Here is my style:
https://github.com/Infixo/Civ6-Mods/blob/master/RET/RealEraTracker.lua
https://github.com/Infixo/Civ6-Mods/blob/master/REU/Lua/RealEurekas.lua
https://github.com/Infixo/Real-Natural-Disasters/blob/master/Main/RealNaturalDisasters.lua
If you find this code hard to read then ok. It is a subjective matter, related to your style, experience, expectations, etc. I am going to argue about that.

**Comment by Infixo** [comment]
As for now, please remove me from the team. And to be clear and avoid misunderstanding - like I said, I am willing to contribute, but I value my time. The amount of extra work (=time) that I am willing to put into something is proportional to the size and importance of the topic. Your current approach is too much focused on rules and compliance, creating thus an unproportionally huge amount of overhead which doesn't add value.

**Comment by alexeyOnGitHub** [comment]
code styles are completely arbitrary and there is no inherent reason why one is better than the other. it that were not the case, these discussions would have ended years ago.

let's do this:
1. follow whatever style is currently present in the file, do not reformat it.
2. only use spaces, forget tabs.
3. check how many spaces a majority of the files in this repo uses and stick to that number for all new files. it can be 2, 4 or 66. from the files that I opened this morning, this number appears to be 2, so let's use 2.  the number does not matter. the consistency does.
4. configure your IDE/editor to auto-format **your new files**. do NOT run that formatter on pre-existing files.
5. these are the guidelines and now laws. a PR should be still accepted if the styling is a bit off. **we value progress over perfection.** 

@Infixo  I would be sad to see you leave, but I understand that it is a volunteering work and you don't have to do what you do not like, obviously. if your concern is about styling, this set of rules should address it.

**Comment by bool3max** [comment]
Are there any pointers that I could use to quickly get accustomed to the codebase? I have a bunch of useful changes in mind but I've never written mods for Civ5/6. I do have lua experience. 

Sorry if this is the wrong thread to ask in.

**Comment by Infixo** [comment]
@alexeyOnGitHub We literally agreed on few things to be done differently, just a couple of posts before.
Ad 1. We agreed that it is up to you and if you wanna reformat it back to 4-space-indent, then feel free to do it. If you follow current formatting then it is obvious that you **will never again** go back to 4-spaces as the entire CQUI is now 2-spaces. And 2-spaces is no go for me, pointed it out in the first post, and it is still a no go. "no go" as "I don't want to be forced to use 2-spaces in my code". Also, I can live with existing 2-spaces and a bit of mixture of both.
Ad. 2. Was agreed.
Ad. 3. See p. 1. Now all is 2-spaces, you wanna stuck with it or not? Most of us here seem to think that 4-spaces is a better choice in the long run, a) the code is more readable and b) that is what Firaxis uses.
Ad. 4. Related to p. 1 and 3.
Ad. 5. I will repeat - most of the code is written by Firaxis, and most of changes will be done by Firaxis. This should be the ideal we strive for if we care about consistency.

I wrote a lot of Lua code of my own, dealt with a lot Firaxis code and often deal with other people's mods and code. I am fine with all of that. The situations when the code is really unreadable usually **does not come from 2 vs. 4 spaces dilemma** but from things like naming, weird lua constructs, harcoding stuff, over-complicating things, not knowing API, etc. These are things that PR should be focused on imo.

**Comment by the-m4a** [comment]
@bool3max - Our contributing markdown file in the base of the repo should have some information.  If it's lacking, we will add more.  Basically, there's two repos of importance: 
- civfanatics\CQUI_Community-Edition  (this one)
- [civfanatics\Civ6-UIFiles-Restructured](https://github.com/civfanatics/Civ6-UIFiles-Restructured?organization=civfanatics&organization=civfanatics)

**This Repo:**
Files found in the Assets\Expansion1 and Assets\Expansion2 folders:
- apply to the Rise & Fall and/or Gathering Storm expansions only
- Structure roughly follows the same structure Firaxis uses for the expansions
- Files that _replace_ the Firaxis-authored files share the same name (e.g. citystates.xml)
- Files that _extend_ the Firaxis-authored files are named _originalFileName_CQUI_expansionX.lua_ (where X is 1 or 2, a few exceptions to this)
Files found in the  Assets\Text, Assets\Texture, Assets\UI folders:
- can apply to the Vanilla (base game), R&F, and GS. 
- Structure roughly follows the same structure Firaxis uses for the expansions
- Files that _replace_ the Firaxis-authored files share the same name (e.g. citystates.xml)
- Files that _extend_ the Firaxis-authored files are named _originalFileName_CQUI_basegame.lua_ (typically)
- Files that contain logic/objects that are used by both Vanilla and Expansions are named _originalFileName_CQUI.lua_ (again, may be some inconsistencies here)

Files found in Integrations are other mods that have been integrated in to CQUI.  
- This is work done by other folks that in the past had agreed to have their mod integrated here.  I am in favor of us discontinuing this practice and allowing these mods to stand-alone (so folks would subscribe to them) and remove the administrative work of keeping them up to date on our end.
- BES = Better Espionage Screen
- BTS = Better Trade Screen
- IDS = Improved Diplomacy Screen (only the Text files are integrated and it's not clear they're used anyway)
- ML = More Lenses
- URS = Updated Report Screen (no longer integrated, but the files remain in our repo for the time being)

The CQUI.modinfo describes the files that are imported.  The SQL files describe sqlite database things.

**Comment by the-m4a** [comment]
So I think the update from Firaxis did _not_ break anything
it seems like it's working

still playing, and since I'm saying this I'm sure I'll find something soon

**Comment by alimulhuda** [comment]
I gave it a quick try too and didn't notice anything that jumped out. I also updated the repo with the unmodded files and when I scanned over those files it seems the changes that are done with files common between the new update files and CQUI were for something called observer mode. Not sure what this observer mode is, but I think if we want to try and break CQUI we would probably have to trigger that mode and see what happens.

**Comment by the-m4a** [comment]
Just FYI - Mod is updated on Steam (I uploaded on 7/2 as well... today's contains all of the things checked in today)

**Comment by wayneb64** [comment]
Has this mod been tested with Frontier Pass yet? 

Is it supported yet or does it only support the latest updates to base plus expansions 1 and 2?

I finally finished my last game and I am planing on going online and letting updates happen on my main PC tomorrow so I will at least have all the updates to the game. I am also considering paying for the new content but not if it has not been tested yet with CQUI.

**Comment by Infixo** [comment]
@wayneb64 
There is no such thing as "Frontier Pass" in the game files, setup, etc.
There is the base game, exp1, exp2 and specific DLCs that add some stuff on top of that.
CQUI works with version 1.0.2.39 of the the base game, exp1 and exp2.

In terms of the DLCs - there are tweaks to support Maya specific housing rules.

**Comment by the-m4a** [comment]
I have new frontier pass, and when they updated last month I hopped on to verify if things were good with it... so should anything spring up from changes via frontier pass or the patches that go with, we should hopefully be quick to respond

@Infixo @alimulhuda @alexeyOnGitHub  - I noticed I can list contributors on the workshop page.  When I went to pick someone, I think it only shows friends.  I wonder if adding any of you as contributors would allow you to also do the workshop updates?

**Comment by Infixo** [comment]
I looked through Google and yeah, people have been complaining for a long time that contributors cannot upload mods. I wasn’t aware. That’s so lame.
But anyway maybe it is worth adding us to have e.g. better visibility in the comments. I actually don’t what contributors can do besides just being listed as ones.

**Comment by wayneb64** [comment]
Is the 2 for water in Housing correct? I thought Maya didn't get a water bonus or is fresh water 2 beyond that?

![image](https://user-images.githubusercontent.com/35745304/87228203-336b3a80-c365-11ea-9712-554ab11c89f5.png)

**Comment by Infixo** [comment]
@wayneb64 Please register your question / issue as a separate issue.

**Comment by wayneb64** [comment]
I don't know if it's an issue or not, so I just asked a question. If it is indeed a bug I will create the issue for it.

**Comment by Infixo** [comment]
Me neither. But if we start solving issues or investigating specific topics, then this is no longer a “general discussion thread”, is it?
“Issue” doesn’t mean that you have an actual bug. It is how things are organized, easier to track, don’t mix up with one another. There is even a label “question” for such cases.

**Comment by wayneb64** [comment]
I did some quick tests on my other PC and a Swedish got 5 water on a river and 2 without the river. A Spanish city got 3 for water on the coast so fresh water adds three and the ocean only adds one. Net net, no issue.

**Comment by chaorace** [comment]
I'm working on updating ML to try and unbreak the minimap on Linux/OSX.

I notice we have a customization in this file, but it doesn't seem to actually have any purpose. I looked over the git history and can see the change originally came from ML, so I have no idea why we marked it as a CQUI change! Is it OK for me to drop this change to the CheckTextureOffset in favor of what ML specifies in the current release?

https://github.com/civfanatics/CQUI_Community-Edition/blob/26fc0d23e9d1a86f179bcba46c3b7546c89c13cd/Integrations/ML/UI/minimappanel.xml#L92

**Comment by the-m4a** [comment]
@chaorace - I think it was marked as CQUI only because it differed from minimappanel.xml in the base files.  So feel free to fix it up!

**Comment by the-m4a** [comment]
I pushed up all the changes since 8/2 to the workshop, huzzah

**Comment by Infixo** [comment]
@the-m4a Would you pls. add me in Steam as author?

**Comment by the-m4a** [comment]
Will do - I wanted to also add @alexeyOnGitHub but last time I looked (a while back) I think our accounts on steam needed to be connected or something?  Will look again later today

**Comment by the-m4a** [comment]
Yeah - when I go to add co-contributors I am limited to only peeps on my friends list, unless there's something I am missing.  Add me (if even temporarily) so I can get you listed

**Comment by alexeyOnGitHub** [comment]
sure, I sent a friend request through the workshop (alskor)

**Comment by the-m4a** [comment]
I didn't see infixo's until I saw yours, alexey, but now you're both listed as pending contributors 👍

**Comment by alimulhuda** [comment]
@the-m4a I just noticed that the mod on Steam workshop has your `.git` folder and `cqui_settings_local.sql` file. So I was thinking you could try using a PowerShell script to automate the download part and always get a clean copy from GitHub.
Something like the stuff below might work
```
$client = new-object System.Net.WebClient
$client.DownloadFile("https://github.com/civfanatics/CQUI_Community-Edition/archive/master.zip", $pwd.ToString() + "\CQUI.zip")
Expand-Archive .\CQUI.zip .
del .\CQUI.zip
```

**Comment by chaorace** [comment]
I would be inclined to just bite the bullet and write two scripts, one batch file for windows and one shell file for OSX/Linux.

Powershell is a huge pain to get working as a standalone script due to MS's default execution policies, even for its home platform. I also don't think it's very realistic to expect most OSX/Linux users to have a PS Core install sitting ready to go.

We would be missing out on PS's handy native unzip function, but I think it's more reasonable to have a git dependency and just clone over the install, as opposed to depending on PS to unpack an archive.

I wonder if there's a "fake git" batch script somewhere on the internet that we could include in the mod package for the sake of cloning? If so, that could work quite well as a sort of polyfill on Windows machines without a git client installed.

**Comment by chaorace** [comment]
Did some more research, no such luck concerning a portable git implementation for Windows. Oh well!

I figure we can use a PS script on windows and chainload that through a batch script to bypass the security policy issues.

For Linux/OSX, we can just provide a shell script with a git dependency, since it's so ubiquitous. 

Doing that would make the process as plug n' play as is possible.

**Comment by the-m4a** [comment]
I am missing something, is this something the folks subscribed to the mod would do?  Or.. for when we're working on it?  Sorry for missing the idea there.

Also.. er... any chance you all could do some code reviews? ;)  (you can download that m4a-ppe branch and run that, it has the changes from the open PRs aside from the one with the Lens stuff)

**Comment by alimulhuda** [comment]
The script was meant to maybe help the-m4a, not for subscribers. I was thinking having a script to help with a clean download from GitHub would help make the process of uploading to Steam Workshop a little less error prone.

**Comment by the-m4a** [comment]
aah understood. :)  When I publish to the steam workshop I just point it at the folder that holds the Mod files, which also happens to be the Git location.

That also explains something - I have set up VSCode to point at a bunch of other mods I have subscribed to, and I noticed that a few would end up under Git control, and I hadn't thought much more of it... but if those mods also have a .git folder, that'd explain why VSCode treats them as such

![image](https://user-images.githubusercontent.com/8787640/91256738-f98ba300-e71c-11ea-9807-252703d69eb2.png)

(MoreLenses, Better Espionage Screen, Better Trade Screen, if you're curious what those 3 are... in total I have pointed them at a bunch, including the DLC folder from Civ, since that has some stuff in there that's potentially useful

![image](https://user-images.githubusercontent.com/8787640/91256842-3788c700-e71d-11ea-9047-f796eb39901f.png)

**Comment by the-m4a** [comment]
Also - I have been limiting the uploads to the workshop to about once a week or so, usually on the weekend, because of the CQUI_Settings_Local override it does (in spite of the recent change that allows that file to exist outside of the mod folder).

**Comment by chaorace** [comment]
My bad! That's what I get for skimming!

**Comment by the-m4a** [comment]
I'm uploading the mod to the steam workshop right now - my internet went out last night and wasn't restored until this morning, which is as super lame as it sounds

**Comment by the-m4a** [comment]
I just noticed that the subscriber count is over 14,000.  Last time I remember looking it was half that.  I know the version published by (...previous person? name escaping me) had an insane subscriber count, but still... 14k when we're not even the top-most "CQUI" in the search results is pretty good!

**Comment by Phillrbanks** [comment]
I'm having an issue with CQUI. The city names don't show and when I meet a new Civ it cause the entire hud to vanish. The only other mod oh have running is Enhanced Mod Manager. I have attached a save file with the issue.

[Bug Report Save.zip](https://github.com/civfanatics/CQUI_Community-Edition/files/5191197/Bug.Report.Save.zip)

**Comment by the-m4a** [comment]
Was playing around tonight with that "Array" item type in the Advanced Settings that they implemented for the Natural Wonders.  Finally aligned all of the stars to at least produce this... which, well, not entirely sure yet if it'll amount to anything, but it may be interesting:

![image](https://user-images.githubusercontent.com/8787640/93174429-ce282280-f6e2-11ea-9977-ec89c3986bdc.png)

**Comment by the-m4a** [comment]
to be clear - this is more about satisfying a curiosity than anything.  The "Array" type that creates this UI isn't all that intuitive... it doesn't seem to allow the settings to be preset, nor does it allow for anything besides true/false values.  Eventually did figure out the icons and whatnot, but ultimately it's just a "ok there's a thing that exists", I guess?

For "fun", in case someone wants to experiment with this stuff
cqui_databaseschema.sql:
```sql
CREATE TABLE 'CQUI_Settings_GameSetup' (
    'Setting' TEXT NOT NULL,
    'Value' INTEGER NOT NULL,
    'Name' TEXT NOT NULL,
    'Domain' TEXT NOT NULL,
    'Icon' TEXT NOT NULL,
    'SortIndex' INTEGER NOT NULL,
    'UxHint' TEXT NOT NULL,
    'Description' TEXT NOT NULL,
    PRIMARY KEY('Setting')
);
```

cqui_settings_gamesetup.sql:
```sql
INSERT OR REPLACE INTO CQUI_Settings_GameSetup -- Don't touch this line!
    VALUES
        ("CQUI_AlwaysOpenTechTrees",             0, "Always Open Tech Trees",              "CQUISettingsGameSetupTrueFalse", "ICON_FEATURE_CLIFFS_DOVER", 0, "SimpleSelectPanel", "Always opens the full tech trees instead of the civic/research picker panels"),
        ("CQUI_AutoapplyArchaeologistLens",      1, "Auto-Apply Archaeologist Lens",       "CQUISettingsGameSetupTrueFalse", "ICON_Gold", 0, "SimpleSelectPanel", "Automatically activates the archaeologist lens when selecting a archaeologist"),
-- etc...
        ("CQUI_TraderShowSortOrder",             0, "Trader Show Sort Order",              "CQUISettingsGameSetupTrueFalse", "ICON_Gold", 0, "SimpleSelectPanel", "Adds a divider between groups in TradeOverview panel");
```

cqui_advancedsetup.xml:
```xml
<GameData>
    <Parameters>
        <Row ConfigurationGroup="Game" ConfigurationId="CQUI_SETTINGS_ARRAY6"                                               ParameterId="CQUISettingsGameSetupQuery6"  Name="[COLOR_Gold]CQUI Settings6[ENDCOLOR]" Description="Click to Set CQUI Settings"      Domain="CQUISettingsGameSetupTrueFalse"    Hash="0" Array="1"  GroupId="AdvancedOptions" SortIndex="506"/>
    </Parameters>
    <DomainValueQueries>
        <Row QueryId="CQUISettingsGameSetupQuery6"/>
    </DomainValueQueries>
    <Queries>
        <Row QueryId="CQUISettingsGameSetupQuery6" SQL="SELECT Domain, Name, Description, Icon, 0 AS Hash, Value AS DefaultValue, Value, SortIndex FROM CQUI_Settings_GameSetup WHERE Domain = 'CQUISettingsGameSetupTrueFalse'"/>
    </Queries>
</GameData>
```

**Comment by the-m4a** [comment]
^^^ that was only 10 days ago?  ugh.  weird with the time issues this month! 
Anyway, I just pushed an update to the Steam workshop after confirming things appeared to be okay with a few short tests.  Haven't had a ton of time in the past week or so for Civ, sadly, so hopefully things continue to be alright!

**Comment by Sorxus** [comment]
To point out new comment on closed thread:

https://github.com/civfanatics/CQUI_Community-Edition/issues/152

**Comment by Infixo** [comment]
Hi there. I might be available for some modding in the coming months again. Am I still a contributor? I don't see the repository on my home screen. BR, Infixo

**Comment by alexeyOnGitHub** [comment]
@Infixo  yes, you are a part of "civfanatics org / civfanatics-cqui team" with WRITE permission for this repo.

**Comment by Bohaska** [comment]
Is this repository dead?

**Comment by the-m4a** [comment]
Not dead, I suppose... I just haven't been paying a ton of attention to it lately.  Work, life, etc pulling in various different directions.  I keep the occasional eye on this and the Steam page to see if it's broken, but for the most part, it's idle?
